packages/core/src/stateUtils.ts

```diff
@@ -1102,21 +1048,64 @@ export function resolveTransition<
   const resolvedStateValue = willTransition
     ? getValue(machine.machine, configuration)
     : undefined;
-  const historyValue = currentState
-    ? currentState.historyValue
-      ? currentState.historyValue
-      : stateTransition.source
-      ? getHistoryValue(machine.machine, currentState.value)
-      : undefined
-    : undefined;
-  const currentContext = currentState ? currentState.context : context;
-  const actions = getActions(
-    machine,
-    stateTransition,
-    currentContext,
-    _event,
-    currentState
+
+  // Entry and exit set
+  const prevConfig = getConfiguration(
+    [],
+    currentState ? getStateNodes(machine, currentState.value) : [machine]
   );
+  const resolvedConfig = stateTransition.configuration.length
+    ? getConfiguration(prevConfig, stateTransition.configuration)
+    : prevConfig;
+
+  for (const sn of resolvedConfig) {
+    if (!has(prevConfig, sn)) {
+      stateTransition.entrySet.push(sn);
+    }
+  }
+  for (const sn of prevConfig) {
+    if (!has(resolvedConfig, sn) || has(stateTransition.exitSet, sn.parent)) {
+      stateTransition.exitSet.push(sn);
+    }
+  }
+
+  if (!stateTransition.source) {
+    stateTransition.exitSet = [];
+
+    // Ensure that root StateNode (machine) is entered
+    stateTransition.entrySet.push(machine);
+  }
+  stateTransition.exitSet.sort((a, b) => b.order - a.order);
+  stateTransition.entrySet.sort((a, b) => a.order - b.order);
+
+  // History
+  const historyMap: Record<
+    string,
+    Array<StateNode<TContext, any, TEvent>>
+  > = currentState ? currentState.historyMap : {};
+  if (currentState && currentState.configuration) {
+    // From SCXML algorithm: https://www.w3.org/TR/scxml/#exitStates
+    for (const exitStateNode of new Set(exitSet)) {
```

Comment:
the `new Set` cast isn't needed here, we can iterate through `exitSet` as is just fine